# Project: Data Wrangling (Part 1)


## Gather

In [1]:
# Import required libaries
import pandas as pd
import numpy as np
import requests
import tweepy
import json
import os
import time

In [2]:
df_1 = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
df_1.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [9]:
# Download the prediction data frame using the request libary 
start = time.time()
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
respond  = requests.get(url, stream = True)
# Open the image_predictions.tsv file and save the respond from the requests.get to it
with open('image_predictions.tsv', mode = 'wb') as file:
    file.write(respond.content)
print('Request is completed. \nIt took {} seconds'.format(time.time() - start))

Request is completed. 
It took 3.286181926727295 seconds


In [4]:
df_2 = pd.read_csv('image_predictions.tsv', sep ='\t')

In [5]:
df_2.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [6]:
# Obtain the tweet data for each tweet_id using tweeter API
# Setting up twitter api
consumer_key = 'CONSUMER_KEY'
consumer_secret = 'CONSUMER_SECRET'
access_token = 'ACCESS_TOKEN'
access_token_secret = 'ACCESS_TOKEN_SECRET'

tweet_ids = df_2.tweet_id

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)
api = tweepy.API(auth, wait_on_rate_limit=True)


count = 0 
fail_dict = {}
start  = time.time()

# open text file to write the requested json file to   
with open('tweet_json.txt', mode = 'w') as file:
    for tweet_id in tweet_ids:
        count += 1 
        print(str(count) + ':' + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode = 'extended')
            print('Success')
# write json file to text flie
            json.dump(tweet._json, file)
# write each json file to a new line
            file.write('\n')
        except tweepy.TweepError as e:
            print('Fail')
            fail_dict['tweet_id'] = e
            pass
end = time.time()
print(end - start)
print(fail_dict)
            
    

1:666020888022790149
Success
2:666029285002620928
Success
3:666033412701032449
Success
4:666044226329800704
Success
5:666049248165822465
Success
6:666050758794694657
Success
7:666051853826850816
Success
8:666055525042405380
Success
9:666057090499244032
Success
10:666058600524156928
Success
11:666063827256086533
Success
12:666071193221509120
Success
13:666073100786774016
Success
14:666082916733198337
Success
15:666094000022159362
Success
16:666099513787052032
Success
17:666102155909144576
Success
18:666104133288665088
Success
19:666268910803644416
Success
20:666273097616637952
Success
21:666287406224695296
Success
22:666293911632134144
Success
23:666337882303524864
Success
24:666345417576210432
Success
25:666353288456101888
Success
26:666362758909284353
Success
27:666373753744588802
Success
28:666396247373291520
Success
29:666407126856765440
Success
30:666411507551481857
Success
31:666418789513326592
Success
32:666421158376562688
Success
33:666428276349472768
Success
34:6664307244263587

Success
270:670815497391357952
Success
271:670822709593571328
Success
272:670823764196741120
Success
273:670826280409919488
Success
274:670832455012716544
Success
275:670833812859932673
Success
276:670838202509447168
Success
277:670840546554966016
Success
278:670842764863651840
Success
279:670995969505435648
Success
280:671109016219725825
Success
281:671115716440031232
Success
282:671122204919246848
Success
283:671134062904504320
Success
284:671138694582165504
Success
285:671141549288370177
Success
286:671147085991960577
Success
287:671151324042559489
Success
288:671154572044468225
Success
289:671159727754231808
Success
290:671163268581498880
Success
291:671166507850801152
Success
292:671182547775299584
Success
293:671186162933985280
Success
294:671347597085433856
Success
295:671355857343524864
Success
296:671357843010908160
Success
297:671362598324076544
Success
298:671390180817915904
Success
299:671485057807351808
Success
300:671486386088865792
Success
301:671488513339211776
Success


Success
535:676936541936185344
Success
536:676942428000112642
Success
537:676946864479084545
Success
538:676948236477857792
Success
539:676949632774234114
Success
540:676957860086095872
Success
541:676975532580409345
Success
542:677187300187611136
Success
543:677228873407442944
Success
544:677269281705472000
Success
545:677301033169788928
Success
546:677314812125323265
Success
547:677328882937298944
Success
548:677331501395156992
Success
549:677334615166730240
Success
550:677530072887205888
Success
551:677547928504967168
Success
552:677557565589463040
Success
553:677565715327688705
Success
554:677573743309385728
Success
555:677644091929329666
Success
556:677662372920729601
Success
557:677673981332312066
Success
558:677687604918272002
Success
559:677698403548192770
Success
560:677700003327029250
Success
561:677716515794329600
Success
562:677895101218201600
Success
563:677918531514703872
Success
564:678021115718029313
Success
565:678255464182861824
Success
566:678278586130948096
Success


Success
800:691321916024623104
Success
801:691416866452082688
Success
802:691444869282295808
Success
803:691459709405118465
Success
804:691483041324204033
Success
805:691675652215414786
Success
806:691756958957883396
Success
807:691820333922455552
Success
808:692017291282812928
Success
809:692142790915014657
Success
810:692158366030913536
Success
811:692187005137076224
Success
812:692417313023332352
Success
813:692530551048294401
Success
814:692535307825213440
Success
815:692568918515392513
Success
816:692752401762250755
Success
817:692828166163931137
Success
818:692894228850999298
Success
819:692901601640583168
Success
820:692905862751522816
Success
821:692919143163629568
Success
822:693095443459342336
Success
823:693109034023534592
Success
824:693155686491000832
Success
825:693231807727280129
Success
826:693262851218264065
Success
827:693280720173801472
Success
828:693486665285931008
Success
829:693590843962331137
Success
830:693622659251335168
Success
831:693629975228977152
Success


Success
1063:715220193576927233
Success
1064:715342466308784130
Success
1065:715360349751484417
Success
1066:715680795826982913
Success
1067:715696743237730304
Success
1068:715733265223708672
Success
1069:715928423106027520
Success
1070:716080869887381504
Success
1071:716285507865542656
Success
1072:716439118184652801
Success
1073:716791146589110272
Success
1074:716802964044845056
Success
1075:717009362452090881
Success
1076:717047459982213120
Success
1077:717421804990701568
Success
1078:717537687239008257
Success
1079:717790033953034240
Success
1080:717841801130979328
Success
1081:718234618122661888
Success
1082:718246886998687744
Success
1083:718454725339934721
Success
1084:718460005985447936
Success
1085:718540630683709445
Success
1086:718613305783398402
Success
1087:718631497683582976
Success
1088:718939241951195136
Success
1089:718971898235854848
Success
1090:719332531645071360
Success
1091:719339463458033665
Success
1092:719367763014393856
Success
1093:719551379208073216
Success


Success
1320:756275833623502848
Success
1321:756288534030475264
Success
1322:756303284449767430
Success
1323:756526248105566208
Success
1324:756651752796094464
Success
1325:756939218950160384
Success
1326:756998049151549440
Success
1327:757354760399941633
Success
1328:757393109802180609
Success
1329:757400162377592832
Success
1330:757596066325864448
Success
1331:757597904299253760
Success
1332:757611664640446465
Success
1333:757725642876129280
Success
1334:757729163776290825
Success
1335:757741869644341248
Success
1336:758041019896193024
Success
1337:758355060040593408
Success
1338:758405701903519748
Success
1339:758467244762497024
Success
1340:758474966123810816
Success
1341:758740312047005698
Success
1342:758828659922702336
Success
1343:758854675097526272
Success
1344:759047813560868866
Success
1345:759099523532779520
Success
1346:759159934323924993
Success
1347:759197388317847553
Success
1348:759447681597108224
Success
1349:759557299618865152
Success
1350:759566828574212096
Fail
135

Success
1577:796080075804475393
Success
1578:796116448414461957
Success
1579:796149749086875649
Success
1580:796177847564038144
Success
1581:796387464403357696
Success
1582:796484825502875648
Success
1583:796759840936919040
Success
1584:796865951799083009
Success
1585:797236660651966464
Success
1586:797545162159308800
Success
1587:797971864723324932
Success
1588:798209839306514432
Success
1589:798340744599797760
Success
1590:798628517273620480
Success
1591:798644042770751489
Success
1592:798665375516884993
Success
1593:798673117451325440
Success
1594:798694562394996736
Success
1595:798697898615730177
Success
1596:798925684722855936
Success
1597:798933969379225600
Success
1598:799063482566066176
Success
1599:799297110730567681
Success
1600:799422933579902976
Success
1601:799757965289017345
Success
1602:799774291445383169
Success
1603:800018252395122689
Success
1604:800141422401830912
Success
1605:800388270626521089
Success
1606:800443802682937345
Success
1607:800459316964663297
Success


Success
1834:836753516572119041
Success
1835:836989968035819520
Success
1836:837012587749474308
Fail
1837:837110210464448512
Success
1838:837366284874571778
Fail
1839:837471256429613056
Success
1840:837482249356513284
Success
1841:837820167694528512
Success
1842:838083903487373313
Success
1843:838476387338051585
Success
1844:838561493054533637
Success
1845:838916489579200512
Success
1846:838921590096166913
Success
1847:839239871831150596
Success
1848:839290600511926273
Success
1849:839549326359670784
Success
1850:839990271299457024
Success
1851:840268004936019968
Success
1852:840370681858686976
Success
1853:840632337062862849
Success
1854:840696689258311684
Success
1855:841077006473256960
Success
1856:841314665196081154
Success
1857:841439858740625411
Success
1858:841680585030541313
Success
1859:841833993020538882
Success
1860:842115215311396866
Success
1861:842163532590374912
Success
1862:842535590457499648
Success
1863:842765311967449089
Success
1864:842846295480000512
Success
1865:8

In [6]:
# Open tweet_json.txt file and used the some of the data in the json file to from a DataFrame
tweet_dict = []
with open('tweet_json.txt') as file:
# Loop throught the each line of the file
    for files in file:
        json_file = files[:-1]
        data = json.loads(json_file) # Read the file as a json file
# Extract the desired data and save to a dictionary
        id_str = data['id_str']
        favorite_count = data['favorite_count']
        retweet_count = data['retweet_count']
        full_text = data['full_text']
        user = data['user']
        statuses_count = user['statuses_count']
        user_favourites_count = user['favourites_count']
        list_count = user['listed_count']
        friends_count = user['friends_count']
        followers_count = user['followers_count']
        tweet_dict.append({'id_str':id_str, 'favorite_count':favorite_count, 'retweet_count':retweet_count, 'statuses_count':statuses_count, 'user_favourites_count':user_favourites_count, 'list_count':list_count, 'friends_count':friends_count, 'followers_count':followers_count, 'full_text':full_text})    
file.close()     
# Make a pandas DataFrame with the created dictionary
df_3 = pd.DataFrame(tweet_dict) 
df_3.to_csv('Tweet_data.csv', index = False)

In [7]:
df_3.head()

id_str  favorite_count  retweet_count  statuses_count  \
0  666020888022790149            2388            458           12577   
1  666029285002620928             120             42           12577   
2  666033412701032449             111             41           12577   
3  666044226329800704             268            130           12577   
4  666049248165822465              96             40           12577   

   user_favourites_count  list_count  friends_count  followers_count  \
0                 145790        5733             17          8824209   
1                 145790        5733             17          8824209   
2                 145790        5733             17          8824209   
3                 145790        5733             17          8824209   
4                 145790        5733             17          8824209   

                                           full_text  
0  Here we have a Japanese Irish Setter. Lost eye...  
1  This is a western brown Mitsubishi terrier. Up...  
2  Here is a very happy pup. Big fan of well-main...  
3  This is a purebred Piers Morgan. Loves to Netf...  
4  Here we have a 1949 1st generation vulpix. Enj...

id_str:  The string representation of the unique identifier for this user\
favorite_count:  Tweet number of likes\
retweet:  Number of retweet\
statuses_count:  The number of Tweets (including retweets) issued by the user\
favorites_count:  The number of tweets this user has liked in the account's lifetime\
listed_count:  The number of public lists that this user is a member of\
friends_count:  The number of users this account is currently following\
followers_count:  The number of followers this account currently has\
full_text: Tweet discription

In [8]:
twitter_a = df_1
predictions = df_2
tweet_a = df_3

## Assess

###### Assessing twitter_a

In [9]:
# Assessing DataFrame visually 
twitter_a.head(20)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0   892420643555336193                    NaN                  NaN   
1   892177421306343426                    NaN                  NaN   
2   891815181378084864                    NaN                  NaN   
3   891689557279858688                    NaN                  NaN   
4   891327558926688256                    NaN                  NaN   
5   891087950875897856                    NaN                  NaN   
6   890971913173991426                    NaN                  NaN   
7   890729181411237888                    NaN                  NaN   
8   890609185150312448                    NaN                  NaN   
9   890240255349198849                    NaN                  NaN   
10  890006608113172480                    NaN                  NaN   
11  889880896479866881                    NaN                  NaN   
12  889665388333682689                    NaN                  NaN   
13  889638837579907072                    NaN                  NaN   
14  889531135344209921                    NaN                  NaN   
15  889278841981685760                    NaN                  NaN   
16  888917238123831296                    NaN                  NaN   
17  888804989199671297                    NaN                  NaN   
18  888554962724278272                    NaN                  NaN   
19  888202515573088257                    NaN                  NaN   

                    timestamp  \
0   2017-08-01 16:23:56 +0000   
1   2017-08-01 00:17:27 +0000   
2   2017-07-31 00:18:03 +0000   
3   2017-07-30 15:58:51 +0000   
4   2017-07-29 16:00:24 +0000   
5   2017-07-29 00:08:17 +0000   
6   2017-07-28 16:27:12 +0000   
7   2017-07-28 00:22:40 +0000   
8   2017-07-27 16:25:51 +0000   
9   2017-07-26 15:59:51 +0000   
10  2017-07-26 00:31:25 +0000   
11  2017-07-25 16:11:53 +0000   
12  2017-07-25 01:55:32 +0000   
13  2017-07-25 00:10:02 +0000   
14  2017-07-24 17:02:04 +0000   
15  2017-07-24 00:19:32 +0000   
16  2017-07-23 00:22:39 +0000   
17  2017-07-22 16:56:37 +0000   
18  2017-07-22 00:23:06 +0000   
19  2017-07-21 01:02:36 +0000   

                                               source  \
0   <a href="http://twitter.com/download/iphone" r...   
1   <a href="http://twitter.com/download/iphone" r...   
2   <a href="http://twitter.com/download/iphone" r...   
3   <a href="http://twitter.com/download/iphone" r...   
4   <a href="http://twitter.com/download/iphone" r...   
5   <a href="http://twitter.com/download/iphone" r...   
6   <a href="http://twitter.com/download/iphone" r...   
7   <a href="http://twitter.com/download/iphone" r...   
8   <a href="http://twitter.com/download/iphone" r...   
9   <a href="http://twitter.com/download/iphone" r...   
10  <a href="http://twitter.com/download/iphone" r...   
11  <a href="http://twitter.com/download/iphone" r...   
12  <a href="http://twitter.com/download/iphone" r...   
13  <a href="http://twitter.com/download/iphone" r...   
14  <a href="http://twitter.com/download/iphone" r...   
15  <a href="http://twitter.com/download/iphone" r...   
16  <a href="http://twitter.com/download/iphone" r...   
17  <a href="http://twitter.com/download/iphone" r...   
18  <a href="http://twitter.com/download/iphone" r...   
19  <a href="http://twitter.com/download/iphone" r...   

                                                 text  retweeted_status_id  \
0   This is Phineas. He's a mystical boy. Only eve...                  NaN   
1   This is Tilly. She's just checking pup on you....                  NaN   
2   This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3   This is Darla. She commenced a snooze mid meal...                  NaN   
4   This is Franklin. He would like you to stop ca...                  NaN   
5   Here we have a majestic great white breaching ...                  NaN   
6   Meet Jax. He enjoys ice cream so much he gets ...                  NaN   
7   When you watch your own

In [10]:
twitter_a.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [11]:
twitter_a[twitter_a['expanded_urls'].isnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
179   857214891891077121           8.571567e+17         1.806710e+08   
185   856330835276025856                    NaN                  NaN   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
274   840698636975636481           8.406983e+17         8.405479e+17   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
346   831926988323639298           8.319030e+17         2.068372e+07   
375   828361771580813312                    NaN                  NaN   
387   826598799820865537           8.265984e+17         4.196984e+09   
409   823333489516937216           8.233264e+17         1.582854e+09   
427   821153421864615936           8.211526e+17         1.132119e+08   
498   813130366689148928           8.131273e+17         4.196984e+09   
513   811647686436880384           8.116272e+17         4.196984e+09   
570   801854953262350336           8.018543e+17         1.185634e+07   
576   800859414831898624           8.008580e+17         2.918590e+08   
611   797165961484890113           7.971238e+17         2.916630e+07   
701   786051337297522688           7.727430e+17         7.305050e+17   
707   785515384317313025                    NaN                  NaN   
843   766714921925144576           7.667118e+17         4.196984e+09   
857   763956972077010945           7.638652e+17         1.584641e+07   
967   750381685133418496           7.501805e+17         4.717297e+09   
1005  747651430853525504           7.476487e+17         4.196984e+09   
1080  738891149612572673           7.384119e+17         3.589728e+08   
1295  707983188426153984           7.079801e+17         2.319108e+09   
1345  704491224099647488           7.044857e+17         2.878549e+07   
1445  696518437233913856                    NaN                  NaN   
1446  696490539101908992           6.964887e+17         4.196984e+09   
1474  693644216740769793           6.936422e+17         4.196984e+09   
1479  693582294167244802           6.935722e+17         1.198989e+09   
1497  692423280028966913           6.924173e+17         4.196984e+09   
1523  690607260360429569           6.903413e+17         4.670367e+08   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         

In [12]:
twitter_a['name'].value_counts()

None       745
a           55
Charlie     12
Lucy        11
Cooper      11
          ... 
Reptar       1
Halo         1
Tobi         1
Crimson      1
DonDon       1
Name: name, Length: 957, dtype: int64

In [13]:
# Checking the uniqueness of each column
twitter_a.nunique()

tweet_id                      2356
in_reply_to_status_id           77
in_reply_to_user_id             31
timestamp                     2356
source                           4
text                          2356
retweeted_status_id            181
retweeted_status_user_id        25
retweeted_status_timestamp     181
expanded_urls                 2218
rating_numerator                40
rating_denominator              18
name                           957
doggo                            2
floofer                          2
pupper                           2
puppo                            2
dtype: int64

In [14]:
# Assess DataFrame to check for duplicates
sum(twitter_a.duplicated())

0

In [15]:
# Assest coloumn value
twitter_a.doggo.value_counts()

None     2259
doggo      97
Name: doggo, dtype: int64

In [16]:
# Assest coloumn value
twitter_a.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [17]:
# View column(s) with rating_denominator of 0
twitter_a.query('rating_denominator == 0')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313  835246439529840640           8.352460e+17           26259576.0   

                     timestamp  \
313  2017-02-24 21:54:03 +0000   

                                                source  \
313  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
313  @jonnysun @Lin_Manuel ok jomny I know you're e...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp expanded_urls  \
313                       NaN                        NaN           NaN   

     rating_numerator  rating_denominator  name doggo floofer pupper puppo  
313               960                   0  None  None    None   None  None

###### Assessing predictions

In [18]:
# Assessing DataFrame visually 
predictions

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

In [19]:
predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [20]:
# Checking the uniqueness of each column
predictions.nunique()

tweet_id    2075
jpg_url     2009
img_num        4
p1           378
p1_conf     2006
p1_dog         2
p2           405
p2_conf     2004
p2_dog         2
p3           408
p3_conf     2006
p3_dog         2
dtype: int64

In [21]:
# Assess DataFrame to check for duplicates
sum(predictions.duplicated())

0

###### Assessing tweet_a

In [22]:
# Assessing DataFrame visually 
tweet_a.head()

id_str  favorite_count  retweet_count  statuses_count  \
0  666020888022790149            2388            458           12577   
1  666029285002620928             120             42           12577   
2  666033412701032449             111             41           12577   
3  666044226329800704             268            130           12577   
4  666049248165822465              96             40           12577   

   user_favourites_count  list_count  friends_count  followers_count  \
0                 145790        5733             17          8824209   
1                 145790        5733             17          8824209   
2                 145790        5733             17          8824209   
3                 145790        5733             17          8824209   
4                 145790        5733             17          8824209   

                                           full_text  
0  Here we have a Japanese Irish Setter. Lost eye...  
1  This is a western brown Mitsubishi terrier. Up...  
2  Here is a very happy pup. Big fan of well-main...  
3  This is a purebred Piers Morgan. Loves to Netf...  
4  Here we have a 1949 1st generation vulpix. Enj...

In [23]:
tweet_a.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2059 entries, 0 to 2058
Data columns (total 9 columns):
id_str                   2059 non-null object
favorite_count           2059 non-null int64
retweet_count            2059 non-null int64
statuses_count           2059 non-null int64
user_favourites_count    2059 non-null int64
list_count               2059 non-null int64
friends_count            2059 non-null int64
followers_count          2059 non-null int64
full_text                2059 non-null object
dtypes: int64(7), object(2)
memory usage: 144.9+ KB


In [24]:
# Checking the uniqueness of each column
tweet_a.nunique()

id_str                   2059
favorite_count           1854
retweet_count            1542
statuses_count              1
user_favourites_count       2
list_count                  1
friends_count               1
followers_count            45
full_text                2059
dtype: int64

In [25]:
# Assess DataFrame to check for duplicates
sum(tweet_a.duplicated())

0



### Quality

> #### Twitter_a
- Several tweets were retweet
- Some tweets are replies. They are associated with missing expanded _urls and other data in thesame row
- Tweet_id is an integral not a string 
- Timestamp is a string not a datetime 
- Doggo,floofer, pupper and puppo column represent nulls as None 
- Tweet_id 835246439529840640 rating denominator is 0
- Rating scale is  inconsistent (rating denominator has 18 unique values)

>#### Predictions
- Tweet_id is an integral not a string 
- Several classifications are invalid

>#### Tweet_a
- Column containing tweet id is named as id_str instead of tweet_id
- The following columns: statuses_count, user_favourites_count, list_count, and friends_count; are not necessary since their values is unique to the user (WeRateDogs) 


### Tideness
>- Similar information is contained in three tables
- Like information is contained in four columns (ie. dog stage in doggo, floofer, pupper and puppo column





## Clean

In [26]:
twitter_a_clean = twitter_a.copy()
predictions_clean = predictions.copy()
tweet_a_clean = tweet_a.copy()

### Quality

###### CLEANING TWITTER_A

#### Define
Drop all retweet by removing columns with a retweeted_status_id

#### Code

In [27]:
twitter_a_clean = twitter_a_clean[twitter_a_clean['retweeted_status_id'].isnull()]

#### Test

In [28]:
twitter_a_clean['retweeted_status_id'].value_counts()

Series([], Name: retweeted_status_id, dtype: int64)

#### Define
Remove replies from the dataset

#### Code

In [29]:
twitter_a_clean.drop(twitter_a_clean[twitter_a_clean['expanded_urls'].isnull()].index, inplace = True)

#### Test

In [30]:
# Check for rows with missing expanded_urls
twitter_a_clean[twitter_a_clean['expanded_urls'].isnull()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

#### Define
change tweet_id from integral to string data type

#### Code

In [31]:
# Change column data type to string 
twitter_a_clean.tweet_id = twitter_a_clean.tweet_id.astype(str)

#### Test

In [32]:
twitter_a_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2117 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2117 non-null object
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     2117 non-null object
source                        2117 non-null object
text                          2117 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2117 non-null int64
rating_denominator            2117 non-null int64
name                          2117 non-null object
doggo                         2117 non-null object
floofer                       2117 non-null object
pupper                        2117 non-null object
puppo                         2117 non-null object
dtypes: float64(4), int64(2), object(

#### Define
change timestamp data type to datetime

#### Code

In [33]:
twitter_a_clean.timestamp = pd.to_datetime(twitter_a_clean.timestamp)

#### Test

In [34]:
twitter_a_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2117 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2117 non-null object
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     2117 non-null datetime64[ns, UTC]
source                        2117 non-null object
text                          2117 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2117 non-null int64
rating_denominator            2117 non-null int64
name                          2117 non-null object
doggo                         2117 non-null object
floofer                       2117 non-null object
pupper                        2117 non-null object
puppo                         2117 non-null object
dtypes: datetime64[ns, U

#### Define
Replace all None with numpy nan to enable finding of nulls easier

#### Code

In [35]:
twitter_a_clean.replace('None', np.nan, inplace = True)

#### Test

In [36]:
twitter_a_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2117 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2117 non-null object
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     2117 non-null datetime64[ns, UTC]
source                        2117 non-null object
text                          2117 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null float64
expanded_urls                 2117 non-null object
rating_numerator              2117 non-null int64
rating_denominator            2117 non-null int64
name                          1495 non-null object
doggo                         84 non-null object
floofer                       10 non-null object
pupper                        232 non-null object
puppo                         24 non-null object
dtypes: datetime64[ns, UTC](1)

#### Define
Change row whose denominator rating is 0 to 1000

#### Code

In [37]:
twitter_a_clean['rating_denominator'].replace(0,1000, inplace =True)

#### Test

In [38]:
twitter_a_clean.query('rating_denominator == 0')

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

#### Define
Make the rating denominator consistent by making all rating to be on a scale of 10. This will aid easy comparison of the rating. After making the rating scale consistent, the rating numerator and denominatior scale will be dropped.

#### Code

In [39]:
twitter_a_clean['rating'] = np.around(twitter_a_clean['rating_numerator']/twitter_a_clean['rating_denominator'] * 10)
twitter_a_clean['rating'] = twitter_a_clean['rating'].astype('int64')
twitter_a_clean.drop(['rating_numerator', 'rating_denominator'], axis =1, inplace = True)

#### Test

In [40]:
twitter_a_clean.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                  timestamp  \
0 2017-08-01 16:23:56+00:00   
1 2017-08-01 00:17:27+00:00   
2 2017-07-31 00:18:03+00:00   
3 2017-07-30 15:58:51+00:00   
4 2017-07-29 16:00:24+00:00   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id  retweeted_status_timestamp  \
0                       NaN                         NaN   
1                       NaN                         NaN   
2                       NaN                         NaN   
3                       NaN                         NaN   
4                       NaN                         NaN   

                                       expanded_urls      name doggo floofer  \
0  https://twitter.com/dog_rates/status/892420643...   Phineas   NaN     NaN   
1  https://twitter.com/dog_rates/status/892177421...     Tilly   NaN     NaN   
2  https://twitter.com/dog_rates/status/891815181...    Archie   NaN     NaN   
3  https://twitter.com/dog_rates/status/891689557...     Darla   NaN     NaN   
4  https://twitter.com/dog_rates/status/891327558...  Franklin   NaN     NaN   

  pupper puppo  rating  
0    NaN   NaN      13  
1    NaN   NaN      13  
2    NaN   NaN      12  
3    NaN   NaN      13  
4    NaN   NaN      12

###### CLEANING PREDICTIONS

#### Define
Change tweet_id data type from integral to string

#### Code

In [41]:
# Change column data type to string 
predictions_clean.tweet_id = predictions_clean.tweet_id.astype(str)

#### Test

In [42]:
predictions_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null object
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(1), object(5)
memory usage: 152.1+ KB


###### CLEANING TWEET_A

#### Define
Change id_str column name to tweet_id

#### Code

In [43]:
tweet_a_clean.rename(columns = {'id_str':'tweet_id'}, inplace = True)

#### Test

In [44]:
tweet_a_clean.head(1)

tweet_id  favorite_count  retweet_count  statuses_count  \
0  666020888022790149            2388            458           12577   

   user_favourites_count  list_count  friends_count  followers_count  \
0                 145790        5733             17          8824209   

                                           full_text  
0  Here we have a Japanese Irish Setter. Lost eye...

#### Define
Remove statuses_count, user_favourites_count, list_count, friends_count and full_text from the table

#### Code

In [45]:
'''These informations are unique to  the source of the data, WeRateDog and are similar for most tweet_id
As a result of this, it will have less impact on the data

For the full_text column, its information is already present in the df_1 as text'''

tweet_a_clean.drop(['statuses_count', 'user_favourites_count', 'list_count', 'friends_count', 'full_text'], axis =1, inplace =True)

#### Test

In [46]:
tweet_a_clean.head()

tweet_id  favorite_count  retweet_count  followers_count
0  666020888022790149            2388            458          8824209
1  666029285002620928             120             42          8824209
2  666033412701032449             111             41          8824209
3  666044226329800704             268            130          8824209
4  666049248165822465              96             40          8824209

### Tideness

#### Define
Restructure DataFrames to three table named; achieve_source, twitter_a and prediction table

#### Code

In [47]:
# Drop all sources for the different table
twitter_a_clean.drop([
     'source','in_reply_to_status_id', 'in_reply_to_user_id', 'expanded_urls', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp', 'text'], axis = 1, inplace = True)

predictions_clean.drop(['jpg_url'], axis = 1, inplace = True)

# Combine tables to form a single data frame
clean_data = pd.merge( twitter_a_clean, tweet_a_clean, on='tweet_id', how='inner')
combined_clean_data = pd.merge( clean_data, predictions_clean, on='tweet_id', how='inner')


#### Test
Preview each the cleaned tables

In [48]:
combined_clean_data.head(1)

tweet_id                 timestamp     name doggo floofer pupper  \
0  892420643555336193 2017-08-01 16:23:56+00:00  Phineas   NaN     NaN    NaN   

  puppo  rating  favorite_count  retweet_count  ...  img_num      p1  \
0   NaN      13           35850           7593  ...        1  orange   

    p1_conf  p1_dog     p2   p2_conf  p2_dog      p3  p3_conf  p3_dog  
0  0.097049   False  bagel  0.085851   False  banana  0.07611   False  

[1 rows x 21 columns]

#### Define
Restructure DataFrames to three table named; achieve_source, twitter_a and prediction table

#### Code

In [49]:
combined_clean_data = pd.melt(combined_clean_data, id_vars = ['tweet_id', 'timestamp', 'name','rating', 'favorite_count', 'retweet_count', 'followers_count', 'img_num', 'p1', 'p1_conf', 'p1_dog', 'p2', 'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog'], var_name= 'class', value_name = 'stage' )
combined_clean_data = combined_clean_data.sort_values(by=['stage']).reset_index(drop=True)
combined_clean_data = combined_clean_data.drop_duplicates(subset=['tweet_id']).reset_index(drop=True)
combined_clean_data.drop('class', axis=1, inplace = True)

#### Test

In [50]:
combined_clean_data.head(3)

tweet_id                 timestamp    name  rating  \
0  890240255349198849 2017-07-26 15:59:51+00:00  Cassie      14   
1  757393109802180609 2016-07-25 01:52:43+00:00     NaN      10   
2  758828659922702336 2016-07-29 00:57:05+00:00     NaN      13   

   favorite_count  retweet_count  followers_count  img_num  \
0           29596           6610          8824238        1   
1            5784           1723          8824230        2   
2           11175           3818          8824230        1   

                         p1   p1_conf  p1_dog                        p2  \
0                  Pembroke  0.511319    True                  Cardigan   
1        Labrador_retriever  0.787125    True  Chesapeake_Bay_retriever   
2  Chesapeake_Bay_retriever  0.480048    True                    vizsla   

    p2_conf  p2_dog          p3   p3_conf  p3_dog  stage  
0  0.451038    True   Chihuahua  0.029248    True  doggo  
1  0.112676    True  Rottweiler  0.048039    True  doggo  
2  0.264522    True  Weimaraner  0.121840    True  doggo

In [51]:
combined_clean_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1987 entries, 0 to 1986
Data columns (total 18 columns):
tweet_id           1987 non-null object
timestamp          1987 non-null datetime64[ns, UTC]
name               1441 non-null object
rating             1987 non-null int64
favorite_count     1987 non-null int64
retweet_count      1987 non-null int64
followers_count    1987 non-null int64
img_num            1987 non-null int64
p1                 1987 non-null object
p1_conf            1987 non-null float64
p1_dog             1987 non-null bool
p2                 1987 non-null object
p2_conf            1987 non-null float64
p2_dog             1987 non-null bool
p3                 1987 non-null object
p3_conf            1987 non-null float64
p3_dog             1987 non-null bool
stage              305 non-null object
dtypes: bool(3), datetime64[ns, UTC](1), float64(3), int64(5), object(6)
memory usage: 238.8+ KB


 ## Final DataFrame

In [52]:
# To store all cleaned DataFrame
combined_clean_data.to_csv('twitter_archive_master.csv', index=False)